protocol/bufferedtcpreader.go

```diff
@@ -0,0 +1,124 @@
+package protocol
+
+import (
+	"github.com/google/gopacket/tcpassembly"
+	"io"
+	"github.com/box/memsniff/capture"
+)
+
+const (
+	maxPackets = 1000
+	maxBytes = 4 * 1024
+)
+
+type ReassemblyQueue struct {
+	reassembled []tcpassembly.Reassembly
+	bytes       capture.BlockBuffer
+	cursor      int
+}
+
+func NewReassemblyQueue() ReassemblyQueue {
+	return ReassemblyQueue{
+		bytes: capture.NewBlockBuffer(maxPackets, maxBytes),
+	}
+}
+
+func (q *ReassemblyQueue) Add(reassembly []tcpassembly.Reassembly) int {
+	lenBefore := len(q.reassembled)
+	q.reassembled = append(q.reassembled, reassembly...)
+	for i := lenBefore; i < len(q.reassembled); i++ {
+		// defensive copy
+		err := q.bytes.Append(q.reassembled[i].Bytes)
+		if err != nil {
+			q.reassembled = q.reassembled[:i]
+			return i - lenBefore
+		}
+		q.reassembled[i].Bytes = q.bytes.Block(i)
+	}
+	return len(reassembly)
+}
+
+func (q *ReassemblyQueue) Next() (*tcpassembly.Reassembly, error) {
+	if q.cursor >= len(q.reassembled) {
+		return nil, io.EOF
+	}
+	q.cursor++
+	return &q.reassembled[q.cursor-1], nil
```

Comment:
feels odd to increment first, then decrement for indexing on the next line. would it be worth considering:

next = &q.reassembled[q.cursor]
q.cursor++
return next